library/ssl_tls.c

```diff
@@ -8213,50 +8212,52 @@ static int ssl_tls12_populate_transform(mbedtls_ssl_transform *transform,
         return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
     }
 
-    ssl_mode = mbedtls_ssl_get_mode_from_ciphersuite(
-#if defined(MBEDTLS_SSL_SOME_SUITES_USE_CBC_ETM)
-        encrypt_then_mac,
-#endif /* MBEDTLS_SSL_SOME_SUITES_USE_CBC_ETM */
-        ciphersuite_info);
-
-    if (ssl_mode == MBEDTLS_SSL_MODE_AEAD) {
-        transform->taglen =
-            ciphersuite_info->flags & MBEDTLS_CIPHERSUITE_SHORT_TAG ? 8 : 16;
-    }
-
 #if defined(MBEDTLS_USE_PSA_CRYPTO)
     if ((status = mbedtls_ssl_cipher_to_psa(ciphersuite_info->cipher,
-                                            transform->taglen,
+                                            0,
                                             &alg,
                                             &key_type,
                                             &key_bits)) != PSA_SUCCESS) {
         ret = psa_ssl_status_to_mbedtls(status);
         MBEDTLS_SSL_DEBUG_RET(1, "mbedtls_ssl_cipher_to_psa", ret);
-        goto end;
+        return ret;
     }
+    ssl_mode = mbedtls_ssl_get_base_mode(alg);
 #else
     cipher_info = mbedtls_cipher_info_from_type(ciphersuite_info->cipher);
     if (cipher_info == NULL) {
         MBEDTLS_SSL_DEBUG_MSG(1, ("cipher info for %u not found",
                                   ciphersuite_info->cipher));
         return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
     }
+    ssl_mode = mbedtls_ssl_get_base_mode(
+        mbedtls_cipher_info_get_mode(cipher_info));
 #endif /* MBEDTLS_USE_PSA_CRYPTO */
 
+    if (ssl_mode == MBEDTLS_SSL_MODE_AEAD) {
```

Comment:
I actually think that his version of the code (lines 8214 - 8253) is less readable and does more work to get the same result as before, at the benefit of not calling `mbedtls_ssl_cipher_to_psa` twice. I can approve this PR if this is a performance optimization or if a second reviewer will determine that this code is better this way, but I would like to voice my concerns.

Reply:
@gstrauss - have you done any performance tests? As I believe that this was the reason you added this as an optimization?

Reply:
The reason I submitted this PR was a result of porting hostap to use mbedtls where I needed to get a specific piece of data.  I had to trace through what I considered convoluted code in ssl_tls.c in order to find out how to obtain this data.

No, I do not have specific performance numbers.  My visual analysis is that the cleaner code flow should be faster, and more importantly, should not be appreciably slower.   Any performance improvement would come from avoiding redundant calls or caching effects of doing work in a tighter scope where the results are used, but I am more interested in slightly cleaner code, which may contribute to later code cleanup once PSA is the default in mbedtls 3.x and the older interface is deprecated/removed.

Reply:
BTW, the work done in the current patch is done for PSA_ALG_CCM and is the same work that would be done by calling `mbedtls_ssl_cipher_to_psa()`, but only doing the work for PSA_ALG_CCM and without the `mbedtls_ssl_cipher_to_psa()` `switch()` and without filling in variables that are then not used.

Please let me know whatever you and the second reviewer think is best.

Reply:
This part of the code is a few more lines than the previous, but not only avoids the second call to `mbedtls_ssl_cipher_to_psa()`, but also separates out `MBEDTLS_SSL_SOME_SUITES_USE_CBC_ETM` and whether or not PSA is used, rather than having a preprocessor conditional for an extra arg to a function whether or not `MBEDTLS_SSL_SOME_SUITES_USE_CBC_ETM` is defined.